### ISSUE TITLE
Look for certificates in all "well-known" locations

### ISSUE THREAD

**Issue by probonopd** [curl/curl#11411]
### I did this

* Compiled and statically linked a program that uses libcurl on Alpine Linux
* Ran the program on openSUSE
* It errored, saying `error setting certificate file: /etc/ssl/certs/ca-certificates.crt`

### I expected the following

Works

### Proposed solution

Unfortunately, Linux distributions seem to all think they need to hide the certificates in different places.

All "well-known" possible locations of certificates should be checked, similar to what https://go.dev/src/crypto/x509/root_linux.go is doing.

I worked around this in my application roughly along those lines, but it would be nice if libcurl did this by default:

```
if (getenv("SSL_CERT_FILE") != NULL) {
            if (curl_easy_setopt(handle, CURLOPT_CAINFO, getenv("SSL_CERT_FILE")) == CURLE_OK) {
                fprintf(stderr, "Using certificate file %s\n", getenv("SSL_CERT_FILE"));
            }
        } else {
            curl_easy_setopt(handle, CURLOPT_CAINFO, NULL);
            const char* certFiles[] = {
            "/etc/ssl/certs/ca-certificates.crt",                // Debian/Ubuntu/Gentoo etc.
            "/etc/pki/tls/certs/ca-bundle.crt",                  // Fedora/RHEL 6
            "/etc/ssl/ca-bundle.pem",                            // OpenSUSE
            "/etc/pki/tls/cacert.pem",                           // OpenELEC
            "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem", // CentOS/RHEL 7
            "/etc/ssl/cert.pem",                                 // Alpine Linux
            };
            for (size_t i = 0; i < sizeof(certFiles) / sizeof(certFiles[0]); ++i) {
                if (access(certFiles[i], F_OK) != -1) {
                    if (curl_easy_setopt(handle, CURLOPT_CAINFO, certFiles[i]) == CURLE_OK) {
                        fprintf(stderr, "Using certificate file %s\n", certFiles[i]);
                        break;
                    }
                }
            }
        }

        if (getenv("SSL_CERT_DIR") != NULL) {
            if (curl_easy_setopt(handle, CURLOPT_CAPATH, getenv("SSL_CERT_DIR")) == CURLE_OK) {
                fprintf(stderr, "Using certificate directory %s\n", getenv("SSL_CERT_DIR"));
            }
        } else {
            curl_easy_setopt(handle, CURLOPT_CAPATH, NULL);
            const char* certDirectories[] = {
                "/etc/ssl/certs",               // SLES10/SLES11, https://golang.org/issue/12139
                "/etc/pki/tls/certs",           // Fedora/RHEL
                "/system/etc/security/cacerts", // Android
            };

            for (size_t i = 0; i < sizeof(certDirectories) / sizeof(certDirectories[0]); ++i) {
                if (access(certDirectories[i], F_OK) != -1) {
                    if (curl_easy_setopt(handle, CURLOPT_CAPATH, certDirectories[i]) == CURLE_OK) {
                        fprintf(stderr, "Using certificate directory %s\n", certDirectories[i]);
                        break;
                    }
                }
            }
 }
```

**Comment by dfandrich** [comment]
Are you using the system's libcurl?  Then this is a packaging problem that you should report to Alpine.

**Comment by probonopd** [comment]
> Are you using the system's libcurl?  Then this is a packaging problem that you should report to Alpine.

When I compile a statically linked binary on Alpine, then it works on Alpine.
But want my binary to run also on openSUSE and other distributions.

**Comment by dfandrich** [comment]
libcurl is designed to use the system's source of certificates, which is
generally in one location. If you want to do run-time probing in your
application, you're free to do that using whatever means you feel necessary and
pass in whichever certificate path or bundle you'd like libcurl to use with
CURLOPT_CAPATH or CURLOPT_CAINFO.

**Comment by bagder** [comment]
I concur with @dfandrich. libcurl should not just try to find any random file that seems to exist and use that as trusted CAs. The exact trust should be fixed and known. Application that think otherwise are free to do such a search.

**Comment by TheAssassin** [comment]
I think libcurl should at least expose a *global* setting so that a sane default can be set once per application when initializing libcurl. This would be a huge UX improvement as users that need to use libcurl on multiple distributions with the same binary (and do *not* want to ship their own CA chain, e.g., to support customized host CA bundles) could follow one single pattern: Look up the correct path yourself, initialize curl with that path and the rest of the entire application is going to work.

I previously suggested on IRC to include such a lookup algorithm inside of libcurl. I think we all agreed that cross-platform compatibility is not libcurl's job. I think users can and probably should handle this themselves. It's just annoying to set the option for every single handle.

As a workaround, I'm currently writing a custom C++ wrapper for libcurl's easy interface that wraps single requests. In the constructor, I can, by default, always set the previously looked up path. But this just affects my use of libcurl there. If other library components I want to use do not do this (and I have no way of telling them of custom paths), this won't work.

TL;DR: libcurl might benefit from a global CA chain directory/file setting.

P.S.: I often see people just deactivate CA validation altogether when trying to use a binary on a different host system. Obviously, this is not the solution. A workaround like "run this lookup/init code before you use any other libcurl code" which can be a bit lengthy but is a one-time thing that can be included in `main()` is something users will be way more willing to implement than having to always call `curl_easy_setopt` once or twice for every handle.

**Comment by bagder** [comment]
> I think libcurl should at least expose a global setting

libcurl does not have an established concept of global settings so it would require quite the change to introduce.

> I think we all agreed that cross-platform compatibility is not libcurl's job

I think libcurl has cross-platform compatibility already - partly because it lets each platform's libcurl set its own CA store's path so that the application can just use the default and it will work with the platforms' own specifics without the application having to know.

**Comment by TheAssassin** [comment]
> cross-platform compatibility

I should have used "cross-distribution portability". My point is completely around being unable to use some compiled libcurl across multiple systems with their system-provided CA chains (and possibly their libssl, ...).

**Comment by sschuberth** [comment]
Could a maintainer please re-close this as "won't fix"? That makes it more clear that the feature was not implemented.

**Comment by icing** [comment]
If it helps.
